unity_plugin/unity_src/Assets/Plugins/iOS/AppStoreDelegate.mm

```diff
@@ -233,36 +252,65 @@ - (void)paymentQueue:(SKPaymentQueue *)queue updatedTransactions:(NSArray *)tran
                 [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
                 break;
-                
+
             case SKPaymentTransactionStateRestored:
-                [self storePurchase:transaction.payment.productIdentifier];
-                UnitySendMessage(EventHandler, "OnPurchaseRestored", MakeStringCopy([transaction.originalTransaction.payment.productIdentifier UTF8String]));
+                jsonTransaction = [self convertTransactionToJson:transaction.originalTransaction storeToUserDefaults:true];
+                if ([jsonTransaction  isEqual: @"error"])
+                {
+                    return;
+                }
+
+                UnitySendMessage(EventHandler, "OnPurchaseRestored", MakeStringCopy([jsonTransaction UTF8String]));
                 [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
                 break;
-                
+
             case SKPaymentTransactionStatePurchased:
-                [self storePurchase:transaction.payment.productIdentifier];
-                NSDictionary *requestContents = [NSDictionary dictionaryWithObjectsAndKeys:
-                                                 transaction.payment.productIdentifier, @"sku",
-                                                 [transaction.transactionReceipt base64Encoding], @"receipt",
-                                                 nil];
-                NSError *error;
-                NSData *requestData = [NSJSONSerialization dataWithJSONObject:requestContents
-                                                                      options:0
-                                                                        error:&error];
-                if (!requestData) {
-                    NSLog(@"Got an error while creating the JSON object: %@", error);
+                jsonTransaction = [self convertTransactionToJson:transaction storeToUserDefaults:true];
+                if ([jsonTransaction  isEqual: @"error"])
+                {
                     return;
                 }
-                
-                NSString * jsonString = [[NSString alloc] initWithData:requestData encoding:NSUTF8StringEncoding];
-                UnitySendMessage(EventHandler, "OnPurchaseSucceeded", MakeStringCopy([jsonString UTF8String]));
+
+                UnitySendMessage(EventHandler, "OnPurchaseSucceeded", MakeStringCopy([jsonTransaction UTF8String]));
                 [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
                 break;
         }
     }
 }
 
+- (NSString*)convertTransactionToJson: (SKPaymentTransaction*) transaction storeToUserDefaults:(bool)store
+{
+    //NSURL *receiptURL = [[NSBundle mainBundle] appStoreReceiptURL];
+    //NSData *receipt = [NSData dataWithContentsOfURL:receiptURL];
+    //NSString *receiptBase64 = [receipt base64EncodedStringWithOptions:0];
+
+    NSString *receiptBase64 = [transaction.transactionReceipt base64EncodedStringWithOptions:0];
```

Comment:
```transaction.transactionReceipt``` was deprecated some time ago. But validation flow changed, so for now It still here.
http://stackoverflow.com/questions/19682663/deprecated-transactionreceipt
The newer API the receipt now contains the list of all transactions performed by the user. The documentation clearly outlines what a receipt looks like:
Meaning that if you really, really wanted to, you can iterate through all the items contained in the receipt to validate against each transaction.